### PR TITLE
Add Streaming Payments extension

### LINKED ISSUE #1024: Implement streaming payments
See figma design [here](https://www.figma.com/proto/dCtvv76S8mk5HNApjwmBto/Expenditures?page-id=1195%3A241884&node-id=1195%3A241891&viewport=301%2C48%2C0.16&scaling=min-zoom&starting-point-node-id=1195%3A241891).

A streaming payment is a type of expenditure in which funds are made available continuously to a user, claimed at arbitrary intervals. A streaming payment is defined within a domain, with a recipient, with a start and end date, and an interval denomination (day, week, etc). A streaming payment can include several tokens, with different amounts to be paid out per interval.

If there are insufficient funds in the domain to cover the owed amounts, a user can claim up to the amount available, with the unclaimed balance still owed to the user.

Once a streaming payment has begun, the end date and token values can be adjusted by users with the relevant permissions (in this case, the funding and administration permissions). If a token value is adjusted, any unclaimed funds are claimed prior to the value adjustment going into effect.

### PR DESCRIPTION
**Pull request by kronosapiens**
Closes #1024

Implements Streaming Payments.

Some implementation notes:

- Creating payments & editing the start & stop times require funding & administration permission.
- The start time can be edited as long as the current start time is in the future (cannot edit a begun payment).
- The end time can be edited as long as the new end time is in the future (cannot set the end time to the past).
- Anyone can claim a payment on behalf of someone else.
- Payouts are tracked per-token, e.g. if funds are not available for one token, the balance will be paid out during a future claim.

### CODE REVIEW

**Review by area — COMMENTED** [review]
* Does a token need to be able to be added to a salary?

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:116]: Yes, this was something I discussed explicitly with the product folks.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:221]: Yes, we discussed this on the call -- I argued that changing the amount while having to keep track of some unpaid balance in a different denomination would be too unweildly.

**Comment by kronosapiens** [comment]
Here's an interesting conundrum -- how to handle possible duplicate tokens? The first approach, validating the `tokens` array for a lack of duplicates, would require a sort algorithm or using a mapping and paying storage costs, both of which seem a bit heavy-handed. There is nothing inherently problematic about duplicate tokens, but a wrinkle emerges in which if we calculate the `amountsToClaim` prior to the funding transfers, then we may conclude that funds are available when they are not for a duplicate token. Calculating `amountsToClaim` inside the loop which does the transfers avoids this problem (as the second transfer might get tagged as insufficient funds, even if the first does not), but that breaks the checks-effects-interactions pattern.

I'm leaning towards breaking the checks-effects-interactions pattern as the solution here, since it's a minor style break, the the other options involve (in my mind) excessive solutions.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:172]: A slither request 🤷

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:266]: Exactly

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:324]: I believe the current restriction on `setEndTime` would prevent that behavior -- you couldn't cancel an already cancelled payment, because the new block time would be later than the existing end time.

Comment (area) [reply contracts/extensions/StreamingPayments.sol:324]: You mean 

```
    require(block.timestamp <= _endTime, "streaming-payments-invalid-end-time");
```

this restriction? Or something else I'm blind to? Because `_endTime` there is the function parameter, which is `block.timestamp` for this call, is it not?

Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:324]: Yup, you're right. I was thinking of something different, which was not allowing an endTime to be set if the existing endTime has elapsed (you can't change something which has ended), which would be a new check.

File: contracts/extensions/StreamingPayments.sol
Comment (area) [reply contracts/extensions/StreamingPayments.sol:172]: I think we need to go against slither and remove this one second window where a streaming payment is both not yet started and started. I can't quite see a way to take things off the rails (which is a good sign that the various `max` and `min` and things you have floating around are doing their job) but it's still a recipe for disaster down the line.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:290]: That's fair, but that behavior is reasonable -- if you set the start time to the past, then the payment has begun, and thus can't be changed.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:172]: Fair enough.

File: contracts/extensions/StreamingPayments.sol
Comment (area) [inline contracts/extensions/StreamingPayments.sol:213]: I definitely don't think this event should be emitted if nothing was claimed.

File: contracts/extensions/StreamingPayments.sol
Comment (area) [inline contracts/extensions/StreamingPayments.sol:102]: Shouldn't... some stuff be disabled on deprecation? Probably worth talking to product about this if you haven't already, I can see arguments for and against allowing editing of already existing payments.

**Review by area — COMMENTED** [review]
I understand why interval isn't editable, but should check with product that's acceptable.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:102]: Great point, totally slipped by me.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:32]: I figured frontend might want to show total payouts by token, so this would let them aggregate all those payouts.

Comment (area) [reply contracts/extensions/StreamingPayments.sol:32]: You mean like 'this extension has paid out all this xDai since it was instantiated'? Across all payments?

Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:32]: More like "let's render the sum payment for paymentId/token"

Comment (area) [reply contracts/extensions/StreamingPayments.sol:32]: Surely a query for the paymentId, and then filtering locally will be adequate in that scenario?

Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:32]: Could also do it that way. I'm neutral.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:197]: This is a good idea. My approach would be to keep track of this via a boolean which flips to `true` if there are any non-zero payments during the first loop.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol]: Turns out it's not needed...

File: contracts/extensions/StreamingPayments.sol
Comment (area) [inline contracts/extensions/StreamingPayments.sol:153]: I think unfortunately this needs to be earlier in the the function (i.e. immediately after creation, before this loop) to avoid issues with the graph if we use it with this extension. If you think about it processing events in order, it'd try to update a token on a streaming payment before it had created it.

File: contracts/extensions/StreamingPayments.sol
Comment (area) [reply contracts/extensions/StreamingPayments.sol:203]: Ah, yeah, that's the way to do it.

**Review by area — COMMENTED** [review]
I think this is looking good. The only thing I want to do that I've not done yet is to try and break the maths with some pathological edge cases.

File: contracts/extensions/StreamingPayments.sol
Comment (area) [inline contracts/extensions/StreamingPayments.sol:130]: Should not be able to be set to zero.

**Review by area — COMMENTED** [review]
Added a couple of tests demonstrating maths breaking.

File: test/extensions/voting-rep.js
Comment (area) [reply test/extensions/voting-rep.js:337]: Because the voting contract doesn't fail that way. Wanted to bring this up so we could explicitly consider whether we're okay with this or not.

File: contracts/extensions/StreamingPayments.sol
Comment (area) [reply contracts/extensions/StreamingPayments.sol:178]: Generally speaking, I'm loathe to store the same data in multiple places. This ends up saving gas if you make six claims over the lifetime of the payment, which doesn't seem like it would be uncommon, but still...

**Comment by area** [comment]
(Don't merge yet, still want to think about naming of that variable)

File: test/extensions/voting-rep.js
Comment (kronosapiens) [reply test/extensions/voting-rep.js:337]: Was there a reason we/I did this? Seems not strictly necessary...

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [reply contracts/extensions/StreamingPayments.sol:180]: I'll let you make the call then.

**Review by kronosapiens — COMMENTED** [review]
Looks good, will review again before approving once you're ready

File: test/extensions/voting-rep.js
Comment (area) [reply test/extensions/voting-rep.js:337]: I don't think there was a reason, no. It would be better if it was auth'd though, as in the scenario where someone is deploying it independently, as-is someone could watch for the code being deployed and grief by calling the function.

File: contracts/extensions/StreamingPayments.sol
Comment (kronosapiens) [inline contracts/extensions/StreamingPayments.sol:187]: Should be `amountEntitledFromStart`

**Review by kronosapiens — COMMENTED** [review]
FYI I can't approve my own PR but consider this "approved"

**Review by area — APPROVED** [review]

**Pull request merged**
